Initial Setup: Apache Pulsar Contributor Repository Structure

liangyepianzhou opened the pull request:
- Original project:https://github.com/StevenLuMT/pulsar-java-contrib
- Vote: https://lists.apache.org/thread/td0j8l1c3l93nny0m5smnsdmb91j1n2y
- Discussion: https://lists.apache.org/thread/75y70j6pqw15271b0tq8j637tb02mrwo
### Description
This pull request aims to establish the foundational structure for the Apache Pulsar contributor repository. The main goals of this initial commit are:

- To create a clear and organized repository structure.
- To define contribution guidelines and best practices.

### Changes Made
- **README.md**: Added a comprehensive README file that outlines the purpose of the repository, how to contribute, and the code of conduct.
- **conrtibutionGuides.md**: Created a CONTRIBUTING document detailing the steps for submitting contributions, including coding standards and pull request guidelines.
- **best-pratice-blogs**: Initialized a documentation directory to hold technical documentation and guides for contributors.
- **pull_request_template.md**: Added pull request templates to streamline the contribution process.

### Next Steps
- Review and refine the contribution guidelines.
- Expand the documentation with more detailed guides and examples.
- Implement a review process for contributions.

### Issues Resolved


### Related Pull Requests
- N/A

### Special Notes
- This is a work-in-progress and may require further refinement based on community feedback.
- Please review the changes and provide your valuable input.

### Reviewers

### Labels
- enhancement
- plug-in
- documentation
- best-practice

StevenLuMT reviewed (approved):
LGTM

poorbarcode commented on openProvider.md:
```diff
@@ -0,0 +1,34 @@
+我们可以根据Pulsar
```
Is this file needed?

liangyepianzhou replied:
Oh, this is a draft I wrote when I responded to Pulsar's external interface, and I accidentally committed it. Thanks for pointing out this issue

lhotari commented on .ci/environments/common/update_quarkus.sh:
```diff
@@ -0,0 +1,39 @@
+    -Dproperty=version.io.quarkus \
+    -DnewVersion=${QUARKUS_VERSION} \
+    -DgenerateBackupPoms=false \
+    -Dmaven.wagon.http.ssl.insecure=true \
```
Why is `-Dmaven.wagon.http.ssl.insecure=true` set? It shouldn't be a default.

lhotari commented on .ci/jenkins/Jenkinsfile.setup-branch:
We don't use Jenkins for CI. This file should be removed.

lhotari commented on .ci/jenkins/dsl/jobs.groovy:
We don't use Jenkins for CI. This file should be removed.

lhotari commented on .ci/jenkins/dsl/test.sh:
We don't use Jenkins for CI. This file should be removed.

lhotari commented on LICENSE.txt:
In `apache/pulsar`, this file is called `LICENSE` instead of `LICENSE.txt`. Perhaps we could be consistent here and make this `LICENSE`.

lhotari commented on LICENSE.txt:
```diff
@@ -0,0 +1,215 @@
+====
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+====
```
this seems like an uncommon solution. It doesn't look right. Perhaps this is to make a license checker to pass?

liangyepianzhou replied:
Yes, It is generated by `mvn license:format` to make  license checker pass.

lhotari commented on README.md:
```diff
@@ -1,2 +1,32 @@
+suggestion.  PRs are always welcome and greatly appreciated, but for larger functional changes a pre-coding introduction
+can be helpful to ensure this is the correct place and that active or conflicting efforts don't exist.
+
+Emeritus maintainers:
+- [Steven Lu](https://github.com/StevenLuMT)
+- [Liangyepian Zhou](https://github.com/liangyepianzhou)
+- [Linlin Duan](https://github.com/AuroraTwinkle)
+- [Fenggan Cai](https://github.com/cai152)
+- [Jia Zhai](https://github.com/jiazhai)
```
How can there be "emeritus maintainers" when this project has just been started?

liangyepianzhou replied:
These are examples from the previous examples repository. I will remove them later.

lhotari commented on README.md:
```diff
@@ -1,2 +1,32 @@
+## Contributing
+
+pulsar-java-contrib is actively in development.  If you have some common use cases for plugins, please contact us and we'll be happy to support.
+Please[open an issue](https://github.com/StevenLuMT/pulsar-java-contrib/issues/new/choose) to share your idea or
```
This URL should be updated.

lhotari commented on README.md:
```diff
@@ -1,2 +1,32 @@
+
+Pulsar java contrib is to provide a non-core code maintenance repository to collect plugin implementations, personalized features, experimental features, and best practices from users.
+
+- [Plugin Contribution Guide](conrtibutionGuides.md) lists the core interfaces in Pulsar that can be implemented by contributors, and provides implementation guidelines for each type of interface.
```
there's a typo in the linked file name.

lhotari commented on README.md:
```diff
@@ -1,2 +1,32 @@
+
+- [Personalization Features](customizationFeatures.md) lists the customized features and experimental features that require modification to the Pulsar source code.
+
+- [Best Practices]([best-pratice-blogs](best-pratice-blogs)) lists the best practices for each function summarized by community contributions.
```
there's a type in the link name / link target

lhotari commented on SECURITY.md:
```diff
@@ -0,0 +1,23 @@
+# Security
+
+## Security Policy
+
+This Project follows the ASF [security vulnerability handling process](https://apache.org/security/#vulnerability-handling).
+
+To report a new vulnerability you have discovered, please follow the [ASF security vulnerability reporting process](https://apache.org/security/#reporting-a-vulnerability). To report a vulnerability for Pulsar, contact the [Apache Security Team](https://www.apache.org/security/). When reporting a vulnerability to [security@apache.org](mailto:security@apache.org), you can copy your email to [priva ...
+
+It is the responsibility of the security vulnerability handling project team (Apache Pulsar PMC in most cases) to make public security vulnerability announcements. You can follow announcements on the [users@pulsar.apache.org](mailto:users@pulsar.apache.org) mailing list. For instructions on how to subscribe, please see https://pulsar.apache.org/contact/.
+
+## Security advisories
+
+This section lists known vulnerabilities and their resolutions. We update this list regularly to keep our users informed about the security status of the project.
+
+### 2024
+
+| Vulnerability ID | Description | Severity | Resolution | CVE (if applicable) |
+|-------------------|-------------|----------|------------|----------------------|
+| CVE-YYYY-XXXXX    | Brief description of the vulnerability | High/Medium/Low | Details of the fix and mitigation steps | CVE identifier |
+| ...               | ...         | ...      | ...        | ...                  |
+
+Please check back regularly for updates on known vulnerabilities and their resolutions.
+
```
This should be replaced with similar content as there is in apache/pulsar
```suggestion
# Security Policy

See https://pulsar.apache.org/security/.
```

lhotari commented on best-pratice-blogs/consume-best-practice.md:
there's a typo in the directory name `best-pratice-blogs`.

lhotari commented on build/.mvn/wrapper/MavenWrapperDownloader.java:
This is from an out-dated maven wrapper version. Since then the maven wrapper has been migrated to Apache Maven project. This is using the out-dated `io.takari.maven-wrapper`. Please update this to 
```
    <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-wrapper-plugin</artifactId>
        <version>3.3.2</version>
    </plugin>
```

lhotari commented on build/.mvn/wrapper/maven-wrapper.properties:
these are outdated versions. Use
```
    <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-wrapper-plugin</artifactId>
        <version>3.3.2</version>
    </plugin>
```
and `mvn wrapper:wrapper` to update. It's better to use recent maven version such as 3.9.8

liangyepianzhou replied:
I noticed that there is no version 3.9.8 in the mvnrepository, and 3.8.1 is the most widely used version. You can also use 3.8.1 first.
https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-compiler-plugin

liangyepianzhou replied:
Or use 3.11.0 to keep consistent with pulsar.

lhotari commented on build/optaplanner-distribution/pom.xml:
How needs this? If it's not relevant for CI of the Apache project, please remove.

lhotari commented on build/optaplanner-distribution/src/main/assembly/assembly-optaplanner-quickstarts.xml:
remove if there's no reason

lhotari commented on build/optaplanner-distribution/src/main/assembly/resources/ReadMe.txt:
remove

lhotari commented on build/optaplanner-distribution/src/main/assembly/scripts/runQuickstarts.bat:
remove

lhotari commented on build/optaplanner-distribution/src/main/assembly/scripts/runQuickstarts.sh:
remove

lhotari commented on build/optaplanner-distribution/src/main/assembly/sources.xml:
remove

lhotari reviewed (changes requested):
Please address the comments. There seems to be some optaplanner files in the repository which should be removed.

lhotari commented:
> * Original project:https://github.com/StevenLuMT/pulsar-java-contrib

There might be a copyright issue if you are contributing content created by others, even when the files have an Apache 2.0 license. I wonder if this was discussed on the mailing list. I hope @dave2wave could advise how to handle this.

liangyepianzhou commented:
> > * Original project:https://github.com/StevenLuMT/pulsar-java-contrib
> 
> There might be a copyright issue if you are contributing content created by others, even when the files have an Apache 2.0 license. I wonder if this was discussed on the mailing list. I hope @dave2wave could advise how to handle this.

There is no copyright issue here, the previous repository is a temporary example. There are currently no contributions from other companies in this repository.  @StevenLuMT s introduced the Apache template to create the repository, @AuroraTwinkle contributed the first listener plugin, and I contributed the subsequent parts. All of these were merged into the sample repository in the form of PR. The three of us are from the same team and have reached an agreement. There is no copyright issue.

liangyepianzhou commented:
> Please address the comments. There seems to be some optaplanner files in the repository which should be removed.

Thank you for your suggestion. Initially, we introduced a template to create this repository. There are indeed some unnecessary files in it. I cleaned it up once before, but there are still some left. Thank you for pointing it out. I will clean it up later.

StevenLuMT commented:
> > > * Original project:https://github.com/StevenLuMT/pulsar-java-contrib
> > 
> > 
> > There might be a copyright issue if you are contributing content created by others, even when the files have an Apache 2.0 license. I wonder if this was discussed on the mailing list. I hope @dave2wave could advise how to handle this.
> 
> There is no copyright issue here, the previous repository is a temporary example. There are currently no contributions from other companies in this repository. @StevenLuMT introduced the Apache template to create the repository, @AuroraTwinkle contributed the first listener plugin, and I contributed the subsequent parts. All of these were merged into the sample repository in the form of PR. The three of us are from the same team and have reached an agreement. There is no copyright issue.

I agree

lhotari commented:
> > > * Original project:https://github.com/StevenLuMT/pulsar-java-contrib
> > 
> > 
> > There might be a copyright issue if you are contributing content created by others, even when the files have an Apache 2.0 license. I wonder if this was discussed on the mailing list. I hope @dave2wave could advise how to handle this.
> 
> There is no copyright issue here, the previous repository is a temporary example. There are currently no contributions from other companies in this repository. @StevenLuMT introduced the Apache template to create the repository, @AuroraTwinkle contributed the first listener plugin, and I contributed the subsequent parts. All of these were merged into the sample repository in the form of PR. The three of us are from the same team and have reached an agreement. There is no copyright issue.

I'm not trying to be picky. I believe that ASF has a process for this type of situation. I hope @dave2wave could confirm what is sufficient for this case.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+        <maven-checkstyle-plugin.version>3.1.2</maven-checkstyle-plugin.version>
+        <puppycrawl.checkstyle.version>8.45.1</puppycrawl.checkstyle.version>
+        <spotless-maven-plugin.version>2.18.0</spotless-maven-plugin.version>
+        <maven-compiler-plugin.version>3.11.0</maven-compiler-plugin.version>
```
most recent is 3.13.0

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+    <properties>
+        <lombok.version>1.18.32</lombok.version>
+        <slf4j.version>2.0.13</slf4j.version>
+        <maven.compiler.source>17</maven.compiler.source>
+        <maven.compiler.target>17</maven.compiler.target>
```
It's recommend to use `maven.compiler.release` instead of source and target. https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-release.html

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+        <pulsar.version>3.0.5</pulsar.version>
+        <maven-checkstyle-plugin.version>3.1.2</maven-checkstyle-plugin.version>
+        <puppycrawl.checkstyle.version>8.45.1</puppycrawl.checkstyle.version>
+        <spotless-maven-plugin.version>2.18.0</spotless-maven-plugin.version>
```
2.43.0 is most recent. 2.18.0 is very old.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <pulsar.version>3.0.5</pulsar.version>
+        <maven-checkstyle-plugin.version>3.1.2</maven-checkstyle-plugin.version>
```
3.4.0 is most recent

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+        <maven.compiler.target>17</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <pulsar.version>3.0.5</pulsar.version>
```
```suggestion
        <pulsar.version>3.0.6</pulsar.version>
```

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+                <!-- Check that source files have appropriate license headers -->
+                <groupId>com.mycila</groupId>
+                <artifactId>license-maven-plugin</artifactId>
+                <version>4.1</version>
```
it would be consistent to have this version in properties too. 4.5 is most recent version

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+                    <dependency>
+                        <groupId>com.mycila</groupId>
+                        <artifactId>license-maven-plugin-git</artifactId>
+                        <version>4.1</version>
```
add version to properties

lhotari commented on pulsar-auth-contrib/pom.xml:
```diff
@@ -0,0 +1,30 @@
+    <parent>
+        <groupId>org.apache</groupId>
+        <artifactId>pulsar-java-contrib</artifactId>
+        <version>29</version>
```
which version is this? it seems that project version isn't specified in the parent pom

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,221 @@
+        <version>29</version>
+    </parent>
+
+    <inceptionYear>2024</inceptionYear>
+
+  <artifactId>pulsar-java-contrib</artifactId>
+  <packaging>pom</packaging>
+  <name>Pulsar Java Contrib</name>
```
indent is messy. please reformat the file.

lhotari commented on pom.xml:
project version seems to be missing. I could be `1.0.0-SNAPSHOT` or something else.

liangyepianzhou replied:
The `suppressions.xml` file is used to specify the check style configuration for the repository[0], and it is consistent with the content in Pulsar[1].

[0]https://github.com/apache/pulsar-java-contrib/pull/2/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R162  

[1]https://github.com/apache/pulsar/blob/66e1a06b247cd032872e5ab454ff8c8c6cd98550/pom.xml#L2170

lhotari commented on src/main/resources/suppressions.xml:
```diff
@@ -0,0 +1,82 @@
+    <!-- suppress most all checks expect below-->
+    <suppress checks="^(?!.*(UnusedImports|IllegalImport)).*$" files=".*[\\/]src[\\/]test[\\/].*"/>
+
+    <suppress checks="IllegalImport" files="ProxyServiceStarter.java"/>
+
+    <!-- suppress all checks in the copied code -->
+    <suppress checks=".*" files=".+[\\/]com[\\/]scurrilous[\\/]circe[\\/].+\.java"/>
+
+    <suppress checks=".*" files="MLDataFormats.java"/>
+    <suppress checks=".*" files="BitSetRecyclable.java"/>
+    <!-- pulsar io -->
+    <suppress checks="RegexpSinglelineJava" files="KafkaConnectData.java"/>
+    <suppress checks="LineLength" files="KafkaConnectData.java"/>
+    <suppress checks="RegexpSinglelineJava" files="KafkaConnectSink.java"/>
+    <!-- pulsar io es -->
+    <suppress checks="LineLength" files="ElasticSearchConfig.java"/>
+    <suppress checks="LineLength" files="ElasticSearchClient.java"/>
+    <!-- pulsar io flume -->
+    <suppress checks="RegexpSingleline" files="Application.java"/>
+    <suppress checks="RegexpSinglelineJava" files="AbstractConfigurationProvider.java"/>
+    <suppress checks="RegexpSinglelineJava" files="AbstractZooKeeperConfigurationProvider.java"/>
+    <suppress checks="RegexpSinglelineJava" files="PollingPropertiesFileConfigurationProvider.java"/>
+    <suppress checks="RegexpSinglelineJava" files="EnvVarResolverProperties.java"/>
+    <!-- pulsar io hbase -->
+    <suppress checks="RegexpSinglelineJava" files="HbaseAbstractSink.java"/>
+    <suppress checks="RegexpSinglelineJava" files="HbaseAbstractConfig.java"/>
+    <suppress checks="RegexpSinglelineJava" files="HbaseSinkConfig.java"/>
+    <!-- pulsar io kafka -->
+    <suppress checks="FallThrough" files="KafkaConnectData.java"/>
+    <!-- pulsar io influxdb -->
+    <suppress checks="RegexpSinglelineJava" files="InfluxDBSinkConfig.java"/>
+    <!-- pulsar io kinesis -->
+    <suppress checks="LineLength" files="KinesisSinkConfig.java"/>
+    <suppress checks="LineLength" files="KinesisSink.java"/>
+    <suppress checks=".*" files=".+[\\/]org[\\/]apache[\\/]pulsar[\\/]io[\\/]kinesis[\\/]fbs[\\/].+.java"/>
+    <!-- pulsar io mongo -->
+    <suppress checks="RegexpSinglelineJava" files="MongoSink.java"/>
+    <suppress checks="RegexpSinglelineJava" files="MongoConfig.java"/>
+    <!-- pulsar io rabbitmq -->
+    <suppress checks="RegexpSinglelineJava" files="RabbitMQAbstractConfig.java"/>
```
it looks like there's unnecessary content here. Please cleanup

liangyepianzhou replied:
The suppressions.xml file is used to specify the check style configuration for the repository[0], and it is consistent with the content in Pulsar[1].

[0]https://github.com/apache/pulsar-java-contrib/pull/2/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R162

[1]https://github.com/apache/pulsar/blob/66e1a06b247cd032872e5ab454ff8c8c6cd98550/pom.xml#L2170

lhotari replied:
well these files are about apache/pulsar repository content. It doesn't make sense to keep these when copying this over to this repository. please remove.

dave2wave commented:
> > * Original project:https://github.com/StevenLuMT/pulsar-java-contrib
> 
> There might be a copyright issue if you are contributing content created by others, even when the files have an Apache 2.0 license. I wonder if this was discussed on the mailing list. I hope @dave2wave could advise how to handle this.

When a contributor who has signed an ICLA makes a commit they have already stated that they will only make contributions that they have a right to license.

If someone is contributing a large amount of code then it may be necessary to go through [IP Clearance](https://incubator.apache.org/ip-clearance/). I don't think that this is required here. It was clear to me in the discussion that repository you are referring to was setup as an example for this new repository.

If you see something that you know is licensed under an [unacceptable license](https://www.apache.org/legal/resolved.html) then say something specific.

If you notice something with a copyright from someone else in any individual's contributions then ask them.

lhotari commented on etc/findbugsExclude.xml:
```diff
@@ -0,0 +1,133 @@
+    <Class name="io.openmessaging.benchmark.driver.jms.JMSBenchmarkTransactionProducer" />
+    <Bug pattern="EI_EXPOSE_REP2" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.kafka.KafkaTopicCreator" />
+    <Bug pattern="RV_RETURN_VALUE_IGNORED_BAD_PRACTICE" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.kop.config.PulsarConfig" />
+    <Bug pattern="UWF_UNWRITTEN_PUBLIC_OR_PROTECTED_FIELD" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.pravega.config.PravegaClientConfig" />
+    <Bug pattern="UWF_UNWRITTEN_PUBLIC_OR_PROTECTED_FIELD" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.rabbitmq.RabbitMqBenchmarkProducer" />
+    <Bug pattern="RV_RETURN_VALUE_OF_PUTIFABSENT_IGNORED" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.rabbitmq.RabbitMqBenchmarkConsumer" />
+    <Bug pattern="EI_EXPOSE_REP2" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.rabbitmq.RabbitMqBenchmarkProducer" />
+    <Bug pattern="EI_EXPOSE_REP2" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.bookkeeper.DlogBenchmarkConsumer" />
+    <Bug pattern="EI_EXPOSE_REP2" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.bookkeeper.DlogBenchmarkProducer" />
+    <Bug pattern="EI_EXPOSE_REP2" />
+  </Match>
+  <Match>
+    <Class name="io.openmessaging.benchmark.driver.bookkeeper.stats.StatsLoggerAdaptor" />
+    <Bug pattern="EI_EXPOSE_REP2" />
+  </Match>
+  <Match>
```
please cleanup this file and remove all content that isn't relevant to this project.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,222 @@
+        <maven-compiler-plugin.version>3.13.0</maven-compiler-plugin.version>
+        <maven-wrapper-plugin.version>3.3.2</maven-wrapper-plugin.version>
+        <license-maven-plugin.version>4.5</license-maven-plugin.version>
+        <junit-jupiter.version>5.7.0</junit-jupiter.version>
```
```suggestion
        <junit-jupiter.version>5.11.0</junit-jupiter.version>
```

lhotari commented on pulsar-client-common-contrib/pom.xml:
```diff
@@ -0,0 +1,51 @@
+            <artifactId>pulsar-client-all</artifactId>
+            <version>${pulsar.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.13.1</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>5.7.0</version>
+            <scope>test</scope>
+        </dependency>
```
use dependencyManagement and junit-bom to set the versions in the parent pom.

lhotari commented on pulsar-client-common-contrib/pom.xml:
```diff
@@ -0,0 +1,51 @@
+    <artifactId>pulsar-client-common-contrib</artifactId>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.pulsar</groupId>
+            <artifactId>pulsar-client-all</artifactId>
+            <version>${pulsar.version}</version>
+        </dependency>
```
use dependencyManagement and pulsar-bom to set the versions in parent pom. Please check https://pulsar.apache.org/docs/3.3.x/client-libraries-java-setup/#pulsar-bom for pulsar-bom instructions.

lhotari commented on LICENSE:
```diff
@@ -0,0 +1,215 @@
+====
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+====
```
Why does this file contain a header? This should be removed and added to exclusions if the license tool enforces it.

lhotari commented on README.md:
```diff
@@ -1,2 +1,25 @@
+- [Personalization Features](customizationFeatures.md) lists the customized features and experimental features that require modification to the Pulsar source code.
+
+- [Best Practices]([best-practice-blogs](best-practice-blogs)) lists the best practices for each function summarized by community contributions.
+  - [consume-best-practice.md](best-pratice-blogs%2Fconsume-best-practice.md)
```
typo here:
```suggestion
  - [consume-best-practice.md](best-practice-blogs%2Fconsume-best-practice.md)
```

lhotari commented on README.md:
```diff
@@ -1,2 +1,25 @@
+  - [consume-best-practice.md](best-pratice-blogs%2Fconsume-best-practice.md)
+
+This project follows the terms of **Apache License 2.0**.
+You can format the code by ` mvn spotless:apply` and generate license headers by `mvn license:format`.
```
```suggestion
You can format the code by `mvn spotless:apply` and generate license headers by `mvn license:format`.
```

lhotari commented on README.md:
```diff
@@ -1,2 +1,25 @@
+## Contributing
+
+pulsar-java-contrib is actively in development.  If you have some common use cases for plugins, please contact us and we'll be happy to support.
+Please[open an issue](https://github.com/apache/pulsar-java-contrib/issues/new) to share your idea or
```
```suggestion
Please [open an issue](https://github.com/apache/pulsar-java-contrib/issues/new) to share your idea or
```

lhotari commented on README.md:
```diff
@@ -1,2 +1,25 @@
+suggestion.  PRs are always welcome and greatly appreciated, but for larger functional changes a pre-coding introduction
+can be helpful to ensure this is the correct place and that active or conflicting efforts don't exist.
+
+Learn more about roles in the [community repository](https://github.com/StevenLuMT/pulsar-java-contrib).
```
```suggestion
Learn more about roles in the [community repository](https://github.com/pulsar/pulsar-java-contrib).
```

lhotari commented on best-practice-blogs/consume-best-practice.md:
```diff
@@ -0,0 +1,229 @@
+Key-shared mode. The main difference between Exclusive and Failover is that in Exclusive mode, the consumer is exclusive
+to the entire topic, while in Failover mode, the consumer is exclusive to only one partition. Failover mode allows backup
+consumer connections that are not consumed. The main difference between Shared and Key-shared is whether their dispatch
+strategy is implemented via a key. For more information about subscription type, refer to the [Pulsar website](https://pulsar.apache.org/docs/3.2.x/concepts-messaging/).
```
```suggestion
strategy is implemented via a key. For more information about subscription type, refer to the [Pulsar website](https://pulsar.apache.org/docs/next/concepts-messaging/).
```

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+        </property>
+      </activation>
+    </profile>
+  </profiles>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>${lombok.version}</version>
+        </dependency>
+
```
indents seems to be misaligned. Please reformat the complete file.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+        </dependency>
+    </dependencies>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.slf4j</groupId>
+                <artifactId>slf4j-bom</artifactId>
+                <version>${slf4j.version}</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.apache.pulsar</groupId>
+                <artifactId>pulsar-bom</artifactId>
+                <version>${pulsar.version}</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+
```
Please place `dependencyManagement` block before the project `dependencies` block.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+        </dependencies>
+    </dependencyManagement>
+
+    <properties>
```
please move the properties in the beginning of the file, for example after the name element. It's more readable that way.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>${maven-compiler-plugin.version}</version>
+                <configuration>
+                    <maven.compiler.release>${maven.compiler.release.version}</maven.compiler.release>
```
this doesn't look right. remove this. there are multiple problems here the element name is `release` instead of `maven.compiler.release`. Docs are https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-release.html .

The maven-compiler-plugin.version must be set and a proper version (> 3.6) for this to work. 
(property docs: https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#release)
Many old maven versions default to ancient maven-compiler-plugin versions that don't support maven.compiler.release or the release configuration. Perhaps you faced that issue at some point?

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+    <properties>
+        <lombok.version>1.18.32</lombok.version>
+        <slf4j.version>2.0.13</slf4j.version>
+        <maven.compiler.release.version>17</maven.compiler.release.version>
```
```suggestion
        <maven.compiler.release>17</maven.compiler.release>
```
Please check the docs https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-release.html

lhotari commented on pulsar-client-common-contrib/pom.xml:
```diff
@@ -0,0 +1,50 @@
+        </dependency>
+    </dependencies>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.apache</groupId>
+                <artifactId>pulsar-java-contrib</artifactId>
+                <version>1.0.0-SNAPSHOT</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
```
I don't think that this is the correct solution. pulsar-java-contrib is already a parent pom, what are you trying to achieve?

liangyepianzhou replied:
After configuring pulsar-bom at the beginning, the pulsar version of the sub-project was incorrect. After adding this to the sub-class, it can work normally. It should be caused by my environment problem at that time. It has been deleted.

StevenLuMT commented:
The current title is good,thanks @lhotari 
the purpose of this PR is to create the structure so that other contributors can develop more clearly.

lhotari commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,68 @@
+- *Added integration tests for end-to-end deployment with large payloads (10MB)*
+- *Extended integration test for recovery after broker failure*
+
+### Matching PR in forked repository
+
+PR in forked repository: <!-- ENTER URL HERE -->
+
+<!--
+After opening this PR, the build in apache/pulsar will fail and instructions will
+be provided for opening a PR in the PR author's forked repository.
+
+apache/pulsar pull requests should be first tested in your own fork since the 
+apache/pulsar CI based on GitHub Actions has constrained resources and quota.
+GitHub Actions provides separate quota for pull requests that are executed in 
+a forked repository.
+
+The tests will be run in the forked repository until all PR review comments have
+been handled, the tests pass and the PR is approved by a reviewer.
+-->
```
This isn't relevant in this repository.

liangyepianzhou replied:
Indeed, the current CI resources of this warehouse are completely sufficient. I will remove this part.

lhotari commented on pom.xml:
```diff
@@ -0,0 +1,239 @@
+        <module>pulsar-auth-contrib</module>
+    </modules>
+
+    <profiles>
+        <profile>
+            <id>default</id>
+            <activation>
+                <property>
+                    <name>!productized</name>
+                </property>
+            </activation>
+        </profile>
+
+        <profile>
+            <id>fullProfile</id>
+            <activation>
+                <property>
+                    <name>full</name>
+                </property>
+            </activation>
+        </profile>
+    </profiles>
```
These profiles don't seem to be used, so they can be removed.

lhotari commented on src/main/resources/checkstyle.xml:
When spotless is used, shouldn't checkstyle be removed completely?

lhotari commented on README.md:
```diff
@@ -1,2 +1,25 @@
+
+This project follows the terms of **Apache License 2.0**.
+You can format the code by `mvn spotless:apply` and generate license headers by `mvn license:format`.
+Please note that the code formatted by Spotless may still not meet the formatting requirements. Please run `mvn checkstyle:check` for inspection.
```
This doesn't make much sense. I think it's better to remove checkstyle completely when spotless is used. `spotless:check` should be used to enforce the configured spotless style.
It's necessary to get this right from the beginning, before accepting contributions.

lhotari commented on etc/checkstyle.xml:
remove checkstyle completely when spotless is used. There's also another checkstyle.xml file in the PR so that's also odd if we were to continue to use checkstyle.

lhotari commented on src/main/resources/suppressions.xml:
if spotless is used, checkstyle could be removed.

lhotari reviewed (changes requested):
Main remaining decision is to decide whether spotless and checkstyle should co-exist. 
I don't think that they should, but I could be wrong. Couldn't we just move to spotless, also in CI to enforce the style.

It's worth putting some effort on this before accepting contributions since formatting will impact all files contributed to the repository.
Any advice on this @nicoloboschi @tisonkun ?

liangyepianzhou commented:
> Main remaining decision is to decide whether spotless and checkstyle should co-exist. I don't think that they should, but I could be wrong. Couldn't we just move to spotless, also in CI to enforce the style.
> 
> It's worth putting some effort on this before accepting contributions since formatting will impact all files contributed to the repository. Any advice on this @nicoloboschi @tisonkun ?

Indeed, normally it is enough to just use spotless when creating a new warehouse. But as a Pulsar incubation project. Because some of the functions may be contributed back to pulsar in the future. So I still added checkstyle and kept it consistent with pulsar. Of course, I could be wrong. 😃

lhotari commented:
> Indeed, normally it is enough to just use spotless when creating a new warehouse. But as a Pulsar incubation project. Because some of the functions may be contributed back to pulsar in the future. So I still added checkstyle and kept it consistent with pulsar. Of course, I could be wrong.

yes you are right. We can keep the co-existence of checkstyle. Spotless references Avro as an example and they also use checkstyle and spotless together: https://github.com/apache/avro/blob/8026c8ffe4ef67ab419dba73910636bf2c1a691c/lang/java/pom.xml#L285-L332

lhotari commented:
I noticed that spotless contains a maven pom formatter too. It would be great to format the pom.xml files with it. https://github.com/diffplug/spotless/tree/main/plugin-maven#maven-pom

lhotari reviewed (approved):
LGTM. The spotless maven pom.xml formatting would be a nice-to-have, but that could be done later.
